_posts/2022-05-19-kotlin-context-receivers.markdown

```diff
@@ -0,0 +1,504 @@
+
+val Person.name get() = "$firstName $lastName"
+val Person.age get() = birthday.until(LocalDate.now(), ChronoUnit.YEARS)
+
+fun Person02.greet(): String {
+  // provides extension receiver for "name" extension property explicitly
+  // provides extension receiver for "age" extension property implicitly
+  return "Hello, I'm ${this.name}, I'm $age years old"
+}
+```
+
+Again, the extensions of a class can also be accessed from outside the class or its extensions, by either providing the
+extension receiver implicitly or explicitly. In fact, the code does not differ from accessing members of a class:
+
+```kotlin
+fun main() {
+  val person = Person("John", "Peters", LocalDate.of(1980, 1, 1))
+
+  // provides extension receiver of greet() extension function explicitly:
+  println(person.greet())
+
+  with(person) {
+    // provides extension receiver of greet() extension function implicitly:
+    println(greet())
+  }
+}
+```
+
+### Member Extensions
+
+In Kotlin it is possible to declare properties or functions that are members of a class, and at the same time extensions
+of another class. They share some characteristics of members and extensions but are also different in some regards:
+
+* member extensions have **both** a dispatch receiver **and** an extension receiver
+* the extension receiver can be provided explicitly (qualified syntax) or implicitly
+* the dispatch receiver can only be provided implicitly (no qualified syntax)
+* `this` within a member extension refers to the extension receiver
+* `this@Type` within a member extension refers to the dispatch receiver
+
+The following example demonstrates characteristics of member extensions:
+
+```kotlin
+class Club(val name: String, val members: MutableSet<Person> = mutableSetOf())
+
+data class Person(val name: String) {
+  fun Club.join() {
+    // "this" within a member extension refers to the extension receiver (Club)
+    // "this@Person" refers to the dispatch receiver (Person)
+    this.members.add(this@Person)
+  }
+
+  fun foundNewClub(clubName: String): Club {
+    val club = Club(clubName)
+    
+    // the extension receiver of the "join" member extension is provided explicitly
+    // the dispatch receiver of the "join" member extension is provided implicitly
+    club.join()
+    
+    return club
+  }
+}
+
+fun main() {
+  val person = Person("John")
+  val club = Club("Kotlin User Group")
+
+  with (person) {
+    // the extension receiver of the "join" member extension is provided explicitly
+    // the dispatch receiver of the "join" member extension is provided implicitly
+    club.join()
+  }
+
+  with (person) {
+    with (club) {
+      // both receivers are provided implicitly
+      join()
+    }
+  }
+}
+```
+We can see in above example that member extensions differ from ordinary members or extensions in one important regard.
+Since the syntax does not allow us to provide more than 1 receiver explicitly, we are forced to provide the dispatch
```

Comment:
```suggestion
We can see in above example that member extensions differ from ordinary members or extensions in one important regard:
Since the syntax does not allow us to provide more than 1 receiver explicitly, we are forced to provide the dispatch
```